streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -1819,23 +1827,39 @@ class ZStream[-R, +E, +A](val process: ZManaged[R, E, Pull[R, E, A]]) extends Se
       for {
         scheduleInit  <- schedule.initial.toManaged_
         schedStateRef <- Ref.make(scheduleInit).toManaged_
-        stream = {
-          def repeated: ZStream[R1 with Clock, E, C] = ZStream.unwrap {
-            for {
-              scheduleState <- schedStateRef.get
-              decision      <- schedule.update((), scheduleState)
-              s2 = if (decision.cont)
-                ZStream
-                  .fromEffect(schedStateRef.set(decision.state) *> clock.sleep(decision.delay))
-                  .drain ++ self.map(f) ++ Stream.succeed(g(decision.finish())) ++ repeated
+        switchPull    <- ZManaged.switchable[R1 with Clock, E, ZIO[R1 with Clock, Option[E], C]]
+        currPull      <- switchPull(self.map(f).process).flatMap(as => Ref.make(as)).toManaged_
+        doneRef       <- Ref.make(false).toManaged_
+        pull = {
+          def go: ZIO[R1 with Clock, Option[E], C] =
+            doneRef.get.flatMap { done =>
+              if (done) ZIO.fail(None)
               else
-                ZStream.empty
-            } yield s2
-          }
-          self.map(f) ++ repeated
+                currPull.get.flatten.foldM(
+                  {
+                    case e @ Some(_) => ZIO.fail(e)
+                    case None =>
+                      schedStateRef.get
+                        .flatMap(schedule.update((), _))
+                        .flatMap { decision =>
+                          if (!decision.cont) doneRef.set(true) *> ZIO.fail(None)
+                          else {
+                            val nextPull =
+                              (ZStream.fromEffect(clock.sleep(decision.delay)).drain ++
+                                self.map(f) ++ Stream.succeed(g(decision.finish()))).process
+
+                            switchPull(nextPull).mapError(Some(_)).tap(currPull.set(_)) *>
```

Comment:
Have you noticed that pattern is coming up all the time with `switchable`? I wonder if we could refactor without having to keep `currPull` and `nextPull` refs around.

Reply:
Yeah, there's some convenience to be had here: we could have the `switchable` constructor return something like:
```
trait Switchable[R, E, A] {
  def switch(m: ZManaged[R, E, A]): ZIO[R, E, A]
  def get: ZIO[R, E, A]
}
```

So the `Ref` is embedded in that.

Reply:
Ah it's actually not that simple. We would have to start requiring an initial resource, or `get` becomes `UIO[Option[A]]`. Beyond that, we need to think what happens if `switch` fails to acquire the resource. What would `get` return afterwards? That also pushes us towards `UIO[Option[A]]`. The problem is that sticking an `Option` in there can be a real pain for calling code.